aml-development/frontend
# feat(config): production webpack config

No ticket for this (yet?).

**Description**
DevOps needs a production configuration for the AML 3.0 frontend

**Acceptance Criteria**
- DevOps has the ability to create a production build for the AML 3.0 frontend
- Developers have the ability to create and run a dev build for the AML 3.0 frontend

**How to Test**
Pull this branch
Run `npm install` to get the package `webpack-merge`

_Production Build_
Look at `webpack.prod.js`.  Certain configuration items have `PLACEHOLDER` as the value.  If necessary, you can change those to the same values in `webpack.dev.js` for testing purposes.
Run `npm run build`
Verify that `app.bundle.js` was built appropriately.  After the build command, that file should be in a `dist` folder in the frontend repo.  Open that file and search for something like:
```
module.exports = {"API_URL":"PLACEHOLDER","NODE_ENV":"production","WEBSOCKET_URL":"PLACEHOLDER","USE_WEBSOCKET":true,"APP_TITLE":"AppsMall","METRICS_URL":"//172.16.105.48/metrics/","METRICS_SITE_ID":2};
```
The actual exported object will be whatever your config is in `webpack.prod.js`.

_Production Run_
Run `npm run prod`.  This command runs a dev server with the production configuration so that developers can test it locally.
Go to `http://localhost:8080/#/` and verify that the app is running in production mode (ie using certs for login, no login page, no logout button, etc)

_Dev Run_
Run `npm run dev`.
Go to `http://localhost:8080/#/` and verify that the app is running in development mode (ie login page, logout button in hamburger menu, etc)

_Custom Configuration_
We now use node's `process.env` style of configuration (ie `SETTING_1=value1 SETTING_2=value2 npm run dev`) for overriding the existing configuration when needed.
For example, if you run the app with the command `USE_WEBSOCKET=false npm run dev`, you should see in the dev console that it does not try to connect to the socket.io server, whereas setting that flag to `true` or not setting it at all does.
